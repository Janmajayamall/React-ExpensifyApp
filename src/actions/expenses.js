import uuid from 'uuid';
import database from './../firebase/firebase.js';
import { get } from 'http';

//Actions for Expenses 
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
            } = expenseData;
   
   const expense = {description, note, amount, createdAt};
          
   return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
     dispatch(addExpense({
         id: ref.key,
         ...expense
     }));
   });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id: id
});

export const startRemoveExpense = (id) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense(id));
        });
    }
}

export const editExpense = (id, update)=>({
    type: 'EDIT_EXPENSE',
    id:id,
    update: update
});

export const startEditExpense = (id, update) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update).then(()=>{
            dispatch(editExpense(id, update));
        });
    }
};



export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses: expenses
})

export const startSetExpenses = () => {
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>{
            const expenses = [];

            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};