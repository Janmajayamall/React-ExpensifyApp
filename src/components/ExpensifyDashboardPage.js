import React from 'react';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom';
import {ExpenseList} from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters.js';
import ExpenseSummaryComponent from './ExpenseSummaryComponent.js';
import { connect } from 'react-redux';
import selectExpenses from './../selectors/expenses.js';


export const ExpensifyDashboardPage = (props) =>{

        return (
            <div>
                <ExpenseSummaryComponent expenses = {props.expenses} />
                <ExpenseListFilters/>
                <ExpenseList expenses = {props.expenses}/>
            </div>
        );}


const mapStateToProps = (state, props)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpensifyDashboardPage);