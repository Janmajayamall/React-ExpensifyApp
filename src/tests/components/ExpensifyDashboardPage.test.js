import React from 'react';
import {shallow} from 'enzyme';
import {ExpensifyDashboardPage} from './../../components/ExpensifyDashboardPage.js';
import expenses from './../fixtures/expenses.js';

test('should render dashboard page', ()=>{
    const wrapper = shallow(<ExpensifyDashboardPage expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});