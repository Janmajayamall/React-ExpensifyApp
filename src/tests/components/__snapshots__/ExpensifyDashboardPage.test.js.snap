// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render dashboard page 1`] = `
<div>
  <ExpenseSummaryComponent
    expenses={
      Array [
        Object {
          "amount": 2,
          "createdAt": -345600000,
          "description": "gum",
          "id": 1,
          "note": "",
        },
        Object {
          "amount": 0,
          "createdAt": 0,
          "description": "cell",
          "id": 2,
          "note": "",
        },
        Object {
          "amount": 1,
          "createdAt": 345600000,
          "description": "bell",
          "id": 3,
          "note": "",
        },
      ]
    }
  />
  <Connect(ExpenseListFilters) />
  <ExpenseList
    expenses={
      Array [
        Object {
          "amount": 2,
          "createdAt": -345600000,
          "description": "gum",
          "id": 1,
          "note": "",
        },
        Object {
          "amount": 0,
          "createdAt": 0,
          "description": "cell",
          "id": 2,
          "note": "",
        },
        Object {
          "amount": 1,
          "createdAt": 345600000,
          "description": "bell",
          "id": 3,
          "note": "",
        },
      ]
    }
  />
</div>
`;
